src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java

```diff
@@ -161,27 +162,132 @@ private MessageParams msg(final MessageType sev, final String cont) {
+        for (int i = position.getLine() + 1; i < contents.size(); i++) {
+            String line = contents.get(i).trim();
+            // If an empty line is encountered, assume the trait's target has not yet been written.
+            if (line.equals("")) {
+                return Optional.empty();
+            // Skip comments lines
+            } else if (line.startsWith("//")) {
+            // Skip other traits.
+            } else if (line.startsWith("@")) {
+                // Jump to end of trait.
+                i = getEndOfTrait(i, contents);
+            } else {
+                // Offset the target shape position by accounting for leading whitespace.
+                String originalLine = contents.get(i);
+                int offset = 1;
+                while (originalLine.charAt(offset) == ' ') {
+                    offset++;
+                }
+                return project.getShapeIdFromLocation(documentUri, new Position(i, offset));
+            }
+        }
+        return Optional.empty();
+    }
+
+    // Determine target shape from an apply statement.
+    private Optional<ShapeId> getApplyStatementTarget(String applyStatement, Optional<String> namespace) {
+        SimpleParser parser = new SimpleParser(applyStatement);
+        parser.expect('a');
+        parser.expect('p');
+        parser.expect('p');
+        parser.expect('l');
+        parser.expect('y');
+        parser.ws();
+        String name = ParserUtils.parseShapeId(parser);
+        if (namespace.isPresent()) {
+            return Optional.of(ShapeId.fromParts(namespace.get(), name));
+        }
+        return Optional.empty();
+    }
+
+    // Find the line where the trait ends.
+    private int getEndOfTrait(int lineNumber, List<String> contents) {
+        String line = contents.get(lineNumber);
+        if (line.contains("(")) {
+            if (hasClosingParen(line)) {
+                return lineNumber;
+            }
+            for (int i = lineNumber + 1; i < contents.size(); i++) {
+                String nextLine = contents.get(i).trim();
+                if (hasClosingParen(nextLine)) {
+                    return i;
+                }
+            }
+        }
+        return lineNumber;
+    }
+
+    // Determine if the line has an unquoted closing parenthesis.
+    private boolean hasClosingParen(String line) {
+        boolean quote = false;
+        for (int i = 0; i < line.length(); i++) {
+            char c = line.charAt(i);
+            if (c == '"' && !quote) {
+                quote = true;
+            } else if (c == '"' && quote) {
+                quote = false;
+            }
+
+            if (c == ')' && !quote) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // Work backwards from current position to determine if position is part of a trait.
+    private boolean isTrait(String documentUri, Position position) throws IOException {
```

Comment:
Yes, this would return `false` if the position is within the trait's value, but this was only intended to return `true` for a position within the trait's shapeId. My assumption there was that if a model author has already added an opening parenthesis and started to write the trait's value, they wouldn't really need completions. Additionally, the above token parser only detects potential shapeIds for now.

Reply:
I think supplying trait property completions would be nice to have, given [the `aws.apigateway#integration` trait's 15 fields of which only 3 are required](https://awslabs.github.io/smithy/1.0/spec/aws/amazon-apigateway.html#aws-apigateway-integration-trait).

Changing the name for now is probably sufficient though.